test/extensions/filters/network/thrift_proxy/integration.cc

```diff
@@ -0,0 +1,115 @@
+#include "test/extensions/filters/network/thrift_proxy/integration.h"
+
+#include <algorithm>
+#include <fstream>
+
+#include "test/test_common/environment.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+
+const std::string& PayloadOptions::modeName() const {
+  static const std::string success = "success";
+  static const std::string idl_exception = "idl-exception";
+  static const std::string exception = "exception";
+
+  switch (mode_) {
+  case DriverMode::Success:
+    return success;
+  case DriverMode::IDLException:
+    return idl_exception;
+  case DriverMode::Exception:
+    return exception;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+const std::string& PayloadOptions::transportName() const {
+  static const std::string framed = "framed";
+  static const std::string unframed = "unframed";
+  static const std::string header = "header";
+
+  switch (transport_) {
+  case TransportType::Framed:
+    return framed;
+  case TransportType::Unframed:
+    return unframed;
+  case TransportType::Header:
+    return header;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+const std::string& PayloadOptions::protocolName() const {
+  static const std::string binary = "binary";
+  static const std::string compact = "compact";
+
+  switch (protocol_) {
+  case ProtocolType::Binary:
+    return binary;
+  case ProtocolType::Compact:
+    return compact;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  }
+}
+
+void BaseThriftIntegrationTest::preparePayloads(const PayloadOptions& options,
+                                                Buffer::Instance& request_buffer,
+                                                Buffer::Instance& response_buffer) {
+  std::vector<std::string> args = {
+      TestEnvironment::runfilesPath(
+          "test/extensions/filters/network/thrift_proxy/driver/generate_fixture.sh"),
```

Comment:
Oh wow, I know this is refactored code, but this is quite the elaborate input generation system. How do you guarantee the availability of the `thrift` Python libraries and tooling? I had a quick look, it was unclear how Bazel's hermeticity is respected.

Reply:
The libraries were added to the external dependencies: envoyproxy/envoy@56a047b5064bba8108ec31c2b6a1f94da1711c20

As far as tooling, the generated python bindings are checked in under the test hierarchy so no tools are needed at build or test time. The upshot is that if ever one wanted to change the service or object definitions you'd need to install apache-thrift to regenerate them, but I think this is better than making the whole of apache-thrift a build dependency.

This seemed like the least invasive way to allow Thrift integration tests. I considered running the client.py/server.py rather than generating fixtures, but orchestrating the processes and ports seemed like more trouble than it was worth.